# Required Section

In [1]:
### importing libraries
import pandas as pd
from tqdm import tqdm

In [2]:
### set up SQL database connector
import json

with open('sql_connector.json', 'r') as f:
    sql_connector = json.load(f)

import sqlalchemy

engine = sqlalchemy.create_engine(
    f"postgresql+psycopg2://{sql_connector['USER']}:{sql_connector['ROOT_PASSWORD']}@localhost/{sql_connector['SCHEMA']}"
)
conn = engine.connect()
del sql_connector

In [8]:
### read csv file
df = pd.read_csv('dataset/delivery_orders_march.csv')

### delete table if it exist in PostgreSQL Schema
conn.execute('DROP TABLE IF EXISTS  delivery_orders CASCADE')

### copy data to SQL Database
start_index = 3176000
df.iloc[:start_index].to_sql('delivery_orders', con= conn, index= False)

In [3]:
### define query helper function
def query(q, con= conn): 
    print(f'''===== START OF QUERY ======
    
{q}

=====   END OF QUERY ======''')
    return pd.read_sql(sqlalchemy.text(q), con= conn)
## alternative
# def query(q, con= conn): 
#     print(f'''===== START OF QUERY ======
    
# {q}

# =====   END OF QUERY ======''')
#     return pd.DataFrame(conn.execute(sqlalchemy.text(q)))

In [4]:
### sanity check
query('select * from delivery_orders').tail()

===== START OF QUERY ======
    
select * from delivery_orders

=====   END OF QUERY ======


orderid        pick  1st_deliver_attempt  2nd_deliver_attempt  \
3175995  31370581557125  1586001108         1.586162e+09                  NaN   
3175996  31371228865904  1585823884         1.585880e+09                  NaN   
3175997  31371234359037  1585892251         1.586163e+09                  NaN   
3175998  31371345167246  1585821713         1.585964e+09                  NaN   
3175999  31371348402192  1585802152         1.586143e+09                  NaN   

                                              buyeraddress  \
3175995  Lipata, Vdf basinillo Advect Condado Balintag ...   
3175996  agdao.davao CW1A ,Labangon ROAD.ALB PRUDENTIAL...   
3175997  caimito,Westmont esternon (Leona ,USEP Brgy.La...   
3175998  daraejung Gohotels.ph brgy.comm.qc Rjc 0928440...   
3175999  5376 St.,#1498 #1799 .sucat 81124 regacho.) CA...   

                                             selleraddress  
3175995  biak Binondo.mla aurora Into.11 baby&beyond ka...  
3175996  K-9 #500 Unison seach 0148 #124-126 7-1 Boy's ...  
3175997  Center,Ortigas, DIONISIO DEPOT-B6 dolce cly 18...  
3175998  156 Serendra, floor) 904 Xpost 5TH Delta, #340...  
3175999  Antipolo Asya Mall，Pasilio Bong st.black the r...

In [5]:
### sanity check
query('select * from delivery_orders').shape

===== START OF QUERY ======
    
select * from delivery_orders

=====   END OF QUERY ======


(3176000, 6)

In [9]:
### sanity check
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3176313 entries, 0 to 3176312
Data columns (total 6 columns):
 #   Column               Dtype  
---  ------               -----  
 0   orderid              int64  
 1   pick                 int64  
 2   1st_deliver_attempt  float64
 3   2nd_deliver_attempt  float64
 4   buyeraddress         object 
 5   selleraddress        object 
dtypes: float64(2), int64(2), object(2)
memory usage: 145.4+ MB


# Optional Section

This section of the notebook is optional to execute. We are executing this section because we didn't directly placed all our samples in PostgreSQL (We only placed 3176000 out of 3176312 samples). To avoid executing this section put all of our samples directly into PostgreSQL.

For now, we may have enough computing power to place more than 3 million samples directly in our SQL database. However, they may be cases where we don't have enough computing power to instantly place so many samples directly into our SQL database. To overcome this issue, we can place our samples to the SQL database in chunks.

The code below demonstrates how to place our sample into PostgreSQL with 1 sample per chunk. There is quite a lot of cleaning process needed on string type feature required to do this. This cleaning process will vary between different SQL database.

In [10]:
### sanity check
df.iloc[start_index].to_frame().T

orderid        pick 1st_deliver_attempt 2nd_deliver_attempt  \
3176000  31371594192588  1585809633         1.58598e+09         1.58625e+09   

                                              buyeraddress  \
3176000  COMMUNITY AcaoBauang GMIK Autogroup, SBC Apong...   

                                             selleraddress  
3176000  BASEMENT taiwan pitong Blk.109 2,3,4 Side arko...

In [11]:
### import regex library
import re

### define helper functions
def add_escape_char(text):
    """
    Pre-process text type features in sample
    """
#    text = re.sub("'", "''", text)
    text = text.replace("'", "''")
    text = 'z'*10 + text + 'z'*10
    return text


def add_to_sql(row, table_name= 'delivery_orders', con= conn):
    """
    Add one data sample to the table_name
    """
    value = list(row)
    for ii in range(len(value)):
        if type(value[ii]) == str:
            value[ii] = add_escape_char(value[ii])
    value = str(tuple(value))
#    value = re.sub("nan", "null", value)
    value = value.replace('nan', 'null')
    value = re.sub(r'zzzzzzzzzz\'|zzzzzzzzzz\"|\'zzzzzzzzzz|\"zzzzzzzzzz', "'", value)
#    value = re.sub(r'\'\'', "''", value)
    value = value.replace(r"\'\'", "''")
    error = True
    try:
        query = f'INSERT INTO {table_name} VALUES {value}'
        conn.execute(query)
    except:
        while error:
            try:
                query = sqlalchemy.text(f'INSERT INTO {table_name} VALUES {value}')
                conn.execute(query)
                error = False
            except Exception as err:
                to_escape = re.findall(r'(?<=bind parameter \').*(?=\'\n)', str(err))[0]
#                value = re.sub(to_escape, '\\' + to_escape, value)
                value = value.replace(to_escape, '\\' + to_escape)

In [12]:
### test the helper functions above
add_to_sql(df.iloc[start_index])

### sanity check
query('select * from delivery_orders').tail()

===== START OF QUERY ======
    
select * from delivery_orders

=====   END OF QUERY ======


orderid        pick  1st_deliver_attempt  2nd_deliver_attempt  \
3175996  31371228865904  1585823884         1.585880e+09                  NaN   
3175997  31371234359037  1585892251         1.586163e+09                  NaN   
3175998  31371345167246  1585821713         1.585964e+09                  NaN   
3175999  31371348402192  1585802152         1.586143e+09                  NaN   
3176000  31371594192588  1585809633         1.585975e+09         1.586252e+09   

                                              buyeraddress  \
3175996  agdao.davao CW1A ,Labangon ROAD.ALB PRUDENTIAL...   
3175997  caimito,Westmont esternon (Leona ,USEP Brgy.La...   
3175998  daraejung Gohotels.ph brgy.comm.qc Rjc 0928440...   
3175999  5376 St.,#1498 #1799 .sucat 81124 regacho.) CA...   
3176000  COMMUNITY AcaoBauang GMIK Autogroup, SBC Apong...   

                                             selleraddress  
3175996  K-9 #500 Unison seach 0148 #124-126 7-1 Boy's ...  
3175997  Center,Ortigas, DIONISIO DEPOT-B6 dolce cly 18...  
3175998  156 Serendra, floor) 904 Xpost 5TH Delta, #340...  
3175999  Antipolo Asya Mall，Pasilio Bong st.black the r...  
3176000  BASEMENT taiwan pitong Blk.109 2,3,4 Side arko...

In [13]:
### sanity check
df.shape

(3176313, 6)

In [14]:
### sanity check
query('SELECT * FROM delivery_orders').shape

===== START OF QUERY ======
    
SELECT * FROM delivery_orders

=====   END OF QUERY ======


(3176001, 6)

In [15]:
### adding remaining samples to SQL database
for ii in tqdm(range(start_index + 1, df.shape[0])):
    add_to_sql(df.iloc[ii])

100%|██████████| 312/312 [00:04<00:00, 75.97it/s]


In [16]:
### sanity check
query('SELECT * FROM delivery_orders').shape

===== START OF QUERY ======
    
SELECT * FROM delivery_orders

=====   END OF QUERY ======


(3176313, 6)

end of optional section

# Required Section (Part 2)

## Brief EDA on Dates

In [17]:
query('SELECT min("1st_deliver_attempt") FROM delivery_orders')

===== START OF QUERY ======
    
SELECT min("1st_deliver_attempt") FROM delivery_orders

=====   END OF QUERY ======


min
0  1.583132e+09

In [21]:
q = [f'to_timestamp({func}("{feature}")) AT TIME ZONE \'Asia/Manila\' AS {func}_{feature}' 
     for func in ['min', 'max'] 
     for feature in ['pick', '1st_deliver_attempt', '2nd_deliver_attempt']]
q = ',\n       '.join(q)
q = '''
SELECT {}
FROM delivery_orders
'''.format(q)

query(q).T

===== START OF QUERY ======
    

SELECT to_timestamp(min("pick")) AT TIME ZONE 'Asia/Manila' AS min_pick,
       to_timestamp(min("1st_deliver_attempt")) AT TIME ZONE 'Asia/Manila' AS min_1st_deliver_attempt,
       to_timestamp(min("2nd_deliver_attempt")) AT TIME ZONE 'Asia/Manila' AS min_2nd_deliver_attempt,
       to_timestamp(max("pick")) AT TIME ZONE 'Asia/Manila' AS max_pick,
       to_timestamp(max("1st_deliver_attempt")) AT TIME ZONE 'Asia/Manila' AS max_1st_deliver_attempt,
       to_timestamp(max("2nd_deliver_attempt")) AT TIME ZONE 'Asia/Manila' AS max_2nd_deliver_attempt
FROM delivery_orders


=====   END OF QUERY ======


0
min_pick                2020-03-02 00:12:23
min_1st_deliver_attempt 2020-03-02 14:45:45
min_2nd_deliver_attempt 2020-03-04 00:01:45
max_pick                2020-04-06 21:57:09
max_1st_deliver_attempt 2020-04-07 21:52:18
max_2nd_deliver_attempt 2020-04-08 03:11:09

In [22]:
q = [f'(to_timestamp({func}("{feature}")) AT TIME ZONE \'Asia/Manila\')::date AS {func}_{feature}' 
     for func in ['min', 'max'] 
     for feature in ['pick', '1st_deliver_attempt', '2nd_deliver_attempt']]
q = ',\n       '.join(q)
q = '''
SELECT {}
FROM delivery_orders
'''.format(q)

query(q).T

===== START OF QUERY ======
    

SELECT (to_timestamp(min("pick")) AT TIME ZONE 'Asia/Manila')::date AS min_pick,
       (to_timestamp(min("1st_deliver_attempt")) AT TIME ZONE 'Asia/Manila')::date AS min_1st_deliver_attempt,
       (to_timestamp(min("2nd_deliver_attempt")) AT TIME ZONE 'Asia/Manila')::date AS min_2nd_deliver_attempt,
       (to_timestamp(max("pick")) AT TIME ZONE 'Asia/Manila')::date AS max_pick,
       (to_timestamp(max("1st_deliver_attempt")) AT TIME ZONE 'Asia/Manila')::date AS max_1st_deliver_attempt,
       (to_timestamp(max("2nd_deliver_attempt")) AT TIME ZONE 'Asia/Manila')::date AS max_2nd_deliver_attempt
FROM delivery_orders


=====   END OF QUERY ======


0
min_pick                 2020-03-02
min_1st_deliver_attempt  2020-03-02
min_2nd_deliver_attempt  2020-03-04
max_pick                 2020-04-06
max_1st_deliver_attempt  2020-04-07
max_2nd_deliver_attempt  2020-04-08

## Pre-processing to exclude holiday dates in calculation

In [23]:
### define holiday dates
holidays = [
    '2020-03-08', #(Sunday);
    '2020-03-25', #(Wednesday)/public holiday in Philippines;
    '2020-03-30', #(Monday)/public holiday in Philippines;
    '2020-03-31'  #(Tuesday)/public holiday in Philippines
]

### convert dates to datetime format
holidays = [pd.to_datetime(i) for i in holidays]

### set every Sunday as holiday
for i in range(1,5):
    holidays.append((holidays[0] + pd.offsets.DateOffset(days= 7*i)))
holidays.sort()

In [27]:
### copy holiday data to SQL database
q = '''
DROP TABLE IF EXISTS holidays;

CREATE TABLE holidays (
    date   date
);

INSERT INTO holidays (date) VALUES
\t{t};
'''.format(t= ',\n\t'.join([f"('{text}')" for text in holidays]))

print(f'''===== START OF QUERY ======

{q}

=====   END OF QUERY ======''')
_ = conn.execute(q)

===== START OF QUERY ======


DROP TABLE IF EXISTS holidays;

CREATE TABLE holidays (
    date   date
);

INSERT INTO holidays (date) VALUES
	('2020-03-08 00:00:00'),
	('2020-03-15 00:00:00'),
	('2020-03-22 00:00:00'),
	('2020-03-25 00:00:00'),
	('2020-03-29 00:00:00'),
	('2020-03-30 00:00:00'),
	('2020-03-31 00:00:00'),
	('2020-04-05 00:00:00');


=====   END OF QUERY ======


In [ ]:
##### alternative code for the cell above
# df_holidays = pd.DataFrame(holidays, columns= ['date'])

### delete table if it exist in PostgreSQL Schema
# conn.execute('DROP TABLE IF EXISTS  holidays')


### copy holiday data to SQL database
# df_holidays.to_sql('holidays', con= conn, index= False)
# df_holidays

In [28]:
### sanity check
query('SELECT * FROM holidays')

===== START OF QUERY ======
    
SELECT * FROM holidays

=====   END OF QUERY ======


date
0  2020-03-08
1  2020-03-15
2  2020-03-22
3  2020-03-25
4  2020-03-29
5  2020-03-30
6  2020-03-31
7  2020-04-05

## Preprocessing SLA_matrix

In [30]:
### read SLA_matrix excel file
df_SLA = pd.read_excel('dataset/SLA_matrix.xlsx')
df_SLA

1st Attempt SLA\n(Working Days)    Unnamed: 1  \
0                                                NaN           NaN   
1                                   Origin\n(Seller)  Metro Manila   
2                                                NaN         Luzon   
3                                                NaN       Visayas   
4                                                NaN      Mindanao   
5                                                NaN           NaN   
6  Working Days are defined as Mon - Sat, Excludi...           NaN   
7  SLA calculation begins from the next day after...           NaN   
8  2nd Attempt must be no later than 3 working da...           NaN   

  Destination (Buyer)      Unnamed: 3      Unnamed: 4      Unnamed: 5  
0        Metro Manila           Luzon         Visayas        Mindanao  
1      3 working days  5 working days  7 working days  7 working days  
2      5 working days  5 working days  7 working days  7 working days  
3      7 working days  7 working days  7 working days  7 working days  
4      7 working days  7 working days  7 working days  7 working days  
5                 NaN             NaN             NaN             NaN  
6                 NaN             NaN             NaN             NaN  
7                 NaN             NaN             NaN             NaN  
8                 NaN             NaN             NaN             NaN

In [31]:
### remove unnecessary data
df_SLA = df_SLA.iloc[1: 5, 1:]
df_SLA.columns = ['seller_city', *df_SLA.iloc[:, 0]]
df_SLA

seller_city    Metro Manila           Luzon         Visayas  \
1  Metro Manila  3 working days  5 working days  7 working days   
2         Luzon  5 working days  5 working days  7 working days   
3       Visayas  7 working days  7 working days  7 working days   
4      Mindanao  7 working days  7 working days  7 working days   

         Mindanao  
1  7 working days  
2  7 working days  
3  7 working days  
4  7 working days

In [33]:
### delete table if it exist in PostgreSQL Schema
conn.execute('DROP TABLE IF EXISTS  sla_matrix CASCADE')

### copy sla_matrix to SQL database
df_SLA.to_sql('sla_matrix', con= conn, index= False)

In [34]:
### sanity check
query('SELECT * FROM sla_matrix')

===== START OF QUERY ======
    
SELECT * FROM sla_matrix

=====   END OF QUERY ======


seller_city    Metro Manila           Luzon         Visayas  \
0  Metro Manila  3 working days  5 working days  7 working days   
1         Luzon  5 working days  5 working days  7 working days   
2       Visayas  7 working days  7 working days  7 working days   
3      Mindanao  7 working days  7 working days  7 working days   

         Mindanao  
0  7 working days  
1  7 working days  
2  7 working days  
3  7 working days

## Unpivoting SLA_matrix

In order to work with the SLA_matrix in SQL, the matrix needs to be unpivoted. The unpivoted matrix will have the following columns:
1. seller_city
2. buyer_city
3. expected_delivery_time (days)

In [35]:
### grab city names
cities = list(query('SELECT seller_city FROM sla_matrix').iloc[:, 0])
cities

===== START OF QUERY ======
    
SELECT seller_city FROM sla_matrix

=====   END OF QUERY ======


['Metro Manila', 'Luzon', 'Visayas', 'Mindanao']

In [42]:
### define helper variable
cross_join_lateral_query_value = [f'(\'{city}\', sla_matrix."{city}")' for city in cities]
cross_join_lateral_query_value = ',\n               '.join(cross_join_lateral_query_value)

In [43]:
### define query string
unpivot_query = f'''
SELECT seller_city, 
       t.buyer_city, 
       CAST(TRIM(TRAILING ' working days' FROM t.working_days) AS INT) AS expected_delivery_time
FROM sla_matrix
    CROSS JOIN LATERAL (
        VALUES {cross_join_lateral_query_value}
    ) as t(buyer_city, working_days)
'''

In [44]:
### execute query + sanity check
query(unpivot_query)

===== START OF QUERY ======
    

SELECT seller_city, 
       t.buyer_city, 
       CAST(TRIM(TRAILING ' working days' FROM t.working_days) AS INT) AS expected_delivery_time
FROM sla_matrix
    CROSS JOIN LATERAL (
        VALUES ('Metro Manila', sla_matrix."Metro Manila"),
               ('Luzon', sla_matrix."Luzon"),
               ('Visayas', sla_matrix."Visayas"),
               ('Mindanao', sla_matrix."Mindanao")
    ) as t(buyer_city, working_days)


=====   END OF QUERY ======


seller_city    buyer_city  expected_delivery_time
0   Metro Manila  Metro Manila                       3
1   Metro Manila         Luzon                       5
2   Metro Manila       Visayas                       7
3   Metro Manila      Mindanao                       7
4          Luzon  Metro Manila                       5
5          Luzon         Luzon                       5
6          Luzon       Visayas                       7
7          Luzon      Mindanao                       7
8        Visayas  Metro Manila                       7
9        Visayas         Luzon                       7
10       Visayas       Visayas                       7
11       Visayas      Mindanao                       7
12      Mindanao  Metro Manila                       7
13      Mindanao         Luzon                       7
14      Mindanao       Visayas                       7
15      Mindanao      Mindanao                       7

In [45]:
### save query result into a view in SQL database
conn.execute(f'CREATE OR REPLACE TEMP VIEW sla_temp AS {unpivot_query}')
### sanity check
query('SELECT * FROM sla_temp')

===== START OF QUERY ======
    
SELECT * FROM sla_temp

=====   END OF QUERY ======


seller_city    buyer_city  expected_delivery_time
0   Metro Manila  Metro Manila                       3
1   Metro Manila         Luzon                       5
2   Metro Manila       Visayas                       7
3   Metro Manila      Mindanao                       7
4          Luzon  Metro Manila                       5
5          Luzon         Luzon                       5
6          Luzon       Visayas                       7
7          Luzon      Mindanao                       7
8        Visayas  Metro Manila                       7
9        Visayas         Luzon                       7
10       Visayas       Visayas                       7
11       Visayas      Mindanao                       7
12      Mindanao  Metro Manila                       7
13      Mindanao         Luzon                       7
14      Mindanao       Visayas                       7
15      Mindanao      Mindanao                       7

## Bonus (How to create a pivot table in PostgreSQL)

In [46]:
query('''
/* Install Postgresql additional function for pivoting*/
CREATE EXTENSION IF NOT EXISTS tablefunc;

SELECT * FROM
CROSSTAB(
    'SELECT seller_city,
            buyer_city,
            expected_delivery_time
     FROM sla_temp
     ORDER BY 1,2',
    $$VALUES {text1}$$
) AS pivot_table(
    seller_city VARCHAR,
    {text2}
)
'''.format(text1= ',\n\t     '.join([f"('{city}')" for city in cities]),
           text2= ',\n    '.join([f'"{city}" INT' for city in cities])))

===== START OF QUERY ======
    

/* Install Postgresql additional function for pivoting*/
CREATE EXTENSION IF NOT EXISTS tablefunc;

SELECT * FROM
CROSSTAB(
    'SELECT seller_city,
            buyer_city,
            expected_delivery_time
     FROM sla_temp
     ORDER BY 1,2',
    $$VALUES ('Metro Manila'),
	     ('Luzon'),
	     ('Visayas'),
	     ('Mindanao')$$
) AS pivot_table(
    seller_city VARCHAR,
    "Metro Manila" INT,
    "Luzon" INT,
    "Visayas" INT,
    "Mindanao" INT
)


=====   END OF QUERY ======


seller_city  Metro Manila  Luzon  Visayas  Mindanao
0         Luzon             5      5        7         7
1  Metro Manila             3      5        7         7
2      Mindanao             7      7        7         7
3       Visayas             7      7        7         7

# Glimpse at seller_city and buyer city distribution

In [49]:
query('''
/* Create helper table for future queries*/
CREATE OR REPLACE TEMP VIEW city_temp AS
SELECT orderid,
       (to_timestamp(pick) AT TIME ZONE 'Asia/Manila')::date AS pick,
       (to_timestamp("1st_deliver_attempt") AT TIME ZONE 'Asia/Manila')::date AS _1st_deliver_attempt,
       (to_timestamp("2nd_deliver_attempt") AT TIME ZONE 'Asia/Manila')::date AS _2nd_deliver_attempt,
       CASE 
           WHEN lower(buyeraddress) LIKE '%metro manila' 
               THEN 'Metro Manila'
           WHEN lower(buyeraddress) LIKE '%luzon' 
               THEN 'Luzon'
           WHEN lower(buyeraddress) LIKE '%visayas' 
               THEN 'Visayas'
           WHEN lower(buyeraddress) LIKE '%mindanao' 
               THEN 'Mindanao'
           ELSE null
       END AS buyer_city,
       CASE 
           WHEN lower(selleraddress) LIKE '%metro manila' 
               THEN 'Metro Manila'
           WHEN lower(selleraddress) LIKE '%luzon' 
               THEN 'Luzon'
           WHEN lower(selleraddress) LIKE '%visayas' 
               THEN 'Visayas'
           WHEN lower(selleraddress) LIKE '%mindanao' 
               THEN 'Mindanao'
           ELSE null
       END AS seller_city
FROM delivery_orders;


/* Install Postgresql additional function for pivoting*/
CREATE EXTENSION IF NOT EXISTS tablefunc;

WITH city_distribution_matrix AS (
/* Pivot Table Query*/
SELECT * FROM
CROSSTAB(
    'SELECT seller_city,
            buyer_city,
            count(1)
     FROM city_temp
     GROUP BY 1,2
     ORDER BY 1,2',
    $$VALUES {text1}$$
) AS pivot_table(
    seller_city VARCHAR,
    {text2}
)
)

/* Query for seller_city x buyer_city count matrix*/
SELECT seller_city AS "seller_city/buyer_city",
       {text4},
       {text5} AS "TOTAL"
FROM (SELECT * FROM city_distribution_matrix
UNION ALL
/*Table Total Query*/
SELECT 'TOTAL' AS seller_city,
       {text3}
FROM city_distribution_matrix) AS q
'''.format(text1= ',\n\t     '.join([f"('{city}')" for city in cities]),
           text2= ',\n    '.join([f'"{city}" INT' for city in cities]),
           text3= ',\n       '.join([f'SUM("{city}") AS "{city}"' for city in cities]),
           text4= ',\n       '.join([f'COALESCE("{city}", 0) AS "{city}"' for city in cities]),
           text5= ' + '.join([f'COALESCE("{city}", 0)' for city in cities])))

===== START OF QUERY ======
    

/* Create helper table for future queries*/
CREATE OR REPLACE TEMP VIEW city_temp AS
SELECT orderid,
       (to_timestamp(pick) AT TIME ZONE 'Asia/Manila')::date AS pick,
       (to_timestamp("1st_deliver_attempt") AT TIME ZONE 'Asia/Manila')::date AS _1st_deliver_attempt,
       (to_timestamp("2nd_deliver_attempt") AT TIME ZONE 'Asia/Manila')::date AS _2nd_deliver_attempt,
       CASE 
           WHEN lower(buyeraddress) LIKE '%metro manila' 
               THEN 'Metro Manila'
           WHEN lower(buyeraddress) LIKE '%luzon' 
               THEN 'Luzon'
           WHEN lower(buyeraddress) LIKE '%visayas' 
               THEN 'Visayas'
           WHEN lower(buyeraddress) LIKE '%mindanao' 
               THEN 'Mindanao'
           ELSE null
       END AS buyer_city,
       CASE 
           WHEN lower(selleraddress) LIKE '%metro manila' 
               THEN 'Metro Manila'
           WHEN lower(selleraddress) LIKE '%luzon' 
               THEN 'Luzon'
  

seller_city/buyer_city  Metro Manila   Luzon  Visayas  Mindanao    TOTAL
0                  Luzon             0       1        0         0        1
1           Metro Manila       1560698  804887   479595    331132  3176312
2                  TOTAL       1560698  804888   479595    331132  3176313

## Final Query

In [53]:
### define query string
q = '''
WITH _1st_delivery_time_query AS (
    SELECT city_temp.orderid,
           SUM(CASE WHEN holidays.date IS NULL THEN 0 ELSE 1 END) AS offset_1
    FROM city_temp
    LEFT JOIN holidays ON holidays.date > city_temp.pick 
                      AND holidays.date < city_temp._1st_deliver_attempt
    GROUP BY 1
)
, _2nd_delivery_time_query AS (
    SELECT city_temp.orderid,
           SUM(CASE WHEN holidays.date IS NULL THEN 0 ELSE 1 END) AS offset_2
    FROM city_temp
    LEFT JOIN holidays ON holidays.date > city_temp._1st_deliver_attempt 
                      AND holidays.date < city_temp._2nd_deliver_attempt
    WHERE city_temp._2nd_deliver_attempt IS NOT NULL
    GROUP BY 1
)
, subquery AS (
SELECT *,
       city_temp._1st_deliver_attempt - city_temp.pick - _1st_delivery_time_query.offset_1 AS delivery_time_1st,
       city_temp._2nd_deliver_attempt - city_temp._1st_deliver_attempt - _2nd_delivery_time_query.offset_2 AS delivery_time_2nd
FROM city_temp
JOIN sla_temp USING(seller_city, buyer_city)
JOIN _1st_delivery_time_query USING(orderid)
LEFT JOIN _2nd_delivery_time_query USING(orderid)
)

SELECT orderid,
       CASE 
           WHEN delivery_time_1st > expected_delivery_time
               THEN 1
           WHEN _2nd_deliver_attempt IS NOT NULL
            AND delivery_time_2nd > 3
               THEN 1
           ELSE 0
       END AS is_late
FROM subquery
'''

print(F'''===== START OF QUERY ======

{q}

=====   END OF QUERY ======''')

### execute query and save the result into variable z
z = conn.execute(sqlalchemy.text(q)).fetchall()

===== START OF QUERY ======


WITH _1st_delivery_time_query AS (
    SELECT city_temp.orderid,
           SUM(CASE WHEN holidays.date IS NULL THEN 0 ELSE 1 END) AS offset_1
    FROM city_temp
    LEFT JOIN holidays ON holidays.date > city_temp.pick 
                      AND holidays.date < city_temp._1st_deliver_attempt
    GROUP BY 1
)
, _2nd_delivery_time_query AS (
    SELECT city_temp.orderid,
           SUM(CASE WHEN holidays.date IS NULL THEN 0 ELSE 1 END) AS offset_2
    FROM city_temp
    LEFT JOIN holidays ON holidays.date > city_temp._1st_deliver_attempt 
                      AND holidays.date < city_temp._2nd_deliver_attempt
    WHERE city_temp._2nd_deliver_attempt IS NOT NULL
    GROUP BY 1
)
, subquery AS (
SELECT *,
       city_temp._1st_deliver_attempt - city_temp.pick - _1st_delivery_time_query.offset_1 AS delivery_time_1st,
       city_temp._2nd_deliver_attempt - city_temp._1st_deliver_attempt - _2nd_delivery_time_query.offset_2 AS delivery_time_2nd
FROM city_temp
JO

In [54]:
### sanity check
z[:5]

[(2195009611, 1),
 (2196681179, 0),
 (2197516565, 0),
 (2198744329, 0),
 (2198892174, 1)]

In [23]:
### write query result to csv file
with open('submission.csv', 'w') as writer:
    writer.write('orderid,is_late\n')
    for values in tqdm(z):
        writer.write(f'{values[0]},{values[1]}\n')

100%|██████████| 3176313/3176313 [00:02<00:00, 1216536.14it/s]


In [37]:
### sanity check using open() built-in function
with open('submission.csv', 'r') as reader:
    for ii in range(10):
        ii += 1
        print(f'line({ii}) \tvalues: {reader.readline()}')

line(1) 	values: orderid,is_late

line(2) 	values: 2201985521,0

line(3) 	values: 2204097418,0

line(4) 	values: 2205812937,0

line(5) 	values: 2205855124,0

line(6) 	values: 2205939517,0

line(7) 	values: 2205966081,0

line(8) 	values: 2206095050,0

line(9) 	values: 2206738323,0

line(10) 	values: 2206946092,0



In [38]:
### sanity check using pandas
df_sub = pd.read_csv('submission.csv')
df_sub.head(10)

orderid  is_late
0  2201985521        0
1  2204097418        0
2  2205812937        0
3  2205855124        0
4  2205939517        0
5  2205966081        0
6  2206095050        0
7  2206738323        0
8  2206946092        0
9  2207331041        1

In [39]:
### sanity check on result shape
df_sub.shape

(3176313, 2)